course-content/31_exploring_data_structures.qmd

```diff
@@ -0,0 +1,447 @@
+
+Each line is executed in sequence, so be careful when constructing code that the order in which you modify your dataframe is right. Take the example below, in which we want to find European values for GDP per capita. In it, we first have a line of code that selects the rows that contain "Europe" in the first line, then looks at the column values for "year", "country", and "gdpPercap".
+
+```{python}
+year_country_gdp = df[(df['continent'] == 'Europe')]
+year_country_gdp = year_country_gdp[['year', 'country', 'gdpPercap']]
+year_country_gdp.head(3)
+```
+
+Had we reversed the lines of code, and selected the "year", "country", and "gdpPercap" columns first to put into our new object, we would not have been able to sort for "Europe" in the "continent" column, as the "continent" would have been essentially removed from the dataframe in the step above, so python would have thrown an error. 
+
+:::{.callout-warning icon="false"}
+
+## Challenge 3
+Your director has come back to you and wondered about whether the life expectancy of people changed during the 1970s in Cambodia. Use what you know about selecting rows and sorting data to get the data you need to answer the question. 
+:::
+:::{.callout-note icon="false" collapse="true"}
+
+### Solution to Challenge 3
+
+Using code to select the country Cambodia, then call `sort_values()` method and chain them together. Looking at the data we see that in the 1970s there was a sharp decline in life expectancy in Cambodia. We also see, thankfully, that it has recovered strongly since then. 
+
+```{python}
+challenge3_df = df[(df['country'] == 'Cambodia')].sort_values('year', ascending = False)
+print(challenge3_df)
+```
+
+:::
+
+## Creating new columns of data
+
+Very often we have some data in our dataset that we want to transform to give us additional information. In Excel this is something that is done all the time by creating a formula in a cell that refers to other columns and applies some sort of logic or mathematical expression to it. 
+
+There are different ways you can go about this in pandas. The most straightforward way is to define a new column on the left of the `=` and then reference the existing column and whatever additional conditions you would like on the right. In the example below, the existing population variable “pop” is converted to a value that shows population in millions.
+
+```{python}
+new_cols_df = df
+new_cols_df['pop_millions'] = new_cols_df['pop']/1000000
+new_cols_df.head(2)
+```
+
+So now we can see which countries had the biggest GDP in 2007. Notice that the GDP data is in scientific notation (i.e. the decimal number times x number of zeros), so it's a bit hard to read. If we wanted readers to consume that data we would go ahead and change the data type for it. But for current purposes we'll leave that alone.
```

Comment:
This sentence should probably go underneath the lambda section as that is where you define GDP